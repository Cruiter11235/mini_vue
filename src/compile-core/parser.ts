import { NodeType } from "./ast";
interface Context {
  source: string;
}
enum TagType {
  Start = "start",
  End = "end",
}

/**
 * 字符串解析入口函数
 * @param content
 */
export function baseParse(content: string) {
  // 初始化上下文
  let context = createParseContent(content);
  // 返回ast的root节点
  return createRoot(parseChildren(context, []));
}
function createParseContent(content: string) {
  return {
    source: content,
  };
}
//
function parseChildren(context: Context, ancestors: any[]): any[] {
  let nodes: any[] = [];
  while (!isEnd(context, ancestors)) {
    const source = context.source;
    if (source.startsWith("{{")) {
      const node = parseInterpolation(context);
      nodes.push(node);
    } else if (source[0] == "<") {
      if (/[a-z]/i.test(source[1])) {
        const node = parseElement(context, ancestors);
        nodes.push(node);
      }
    } else {
      const node = parseText(context);
      nodes.push(node);
    }
  }
  return nodes;
}

// 解析插值语句
function parseInterpolation(context: Context) {
  const openDelimiter = "{{";
  const closeDelimiter = "}}";

  // remove begin '{{'
  advanceBy(context, openDelimiter.length);
  const closeIndex = context.source.indexOf(closeDelimiter);
  const rawContentLength = closeIndex;
  // get content and advance
  const content = parseTextData(context, rawContentLength).trim();
  // advance end delimiter
  advanceBy(context, closeDelimiter.length);
  return {
    type: NodeType.INTERPOLATION,
    content: {
      type: NodeType.SIMPLE_EXPRESSION,
      content,
    },
  };
}
/**
 *
 * @param context
 * @param ancestors 开始标签
 * @returns
 */
function parseElement(context: Context, ancestors: any[]) {
  // parse a tag
  const element: any = parseTag(context, TagType.Start); // parse start tag get tagName
  ancestors.push(element);
  element.children = parseChildren(context, ancestors); // parse children to elementNode
  ancestors.pop();
  if (startWithEndTagOpen(context, element)) {
    parseTag(context, TagType.End);
  } else {
    throw new Error("缺少结束标签");
  }
  return element;
}
function startWithEndTagOpen(context: Context, element: any) {
  return context.source.slice(2, 2 + element.tag.length) === element.tag;
}

function parseTag(context: Context, tagType: TagType) {
  const match = /^<\/?([a-z]+)>/i.exec(context.source)!;
  const tag = match[1]; // tagName
  advanceBy(context, match[0].length);
  if (tagType === TagType.End) return; // return if end tag
  return {
    type: NodeType.ELEMENT,
    tag,
  };
}
function parseText(context: Context) {
  let endIndex = context.source.length;
  let endTokens = ["{{", "<"];
  for (let i = 0; i < endTokens.length; i++) {
    const endTokenIndex = context.source.indexOf(endTokens[i]);
    if (endTokenIndex !== -1 && endIndex > endTokenIndex) {
      endIndex = endTokenIndex;
    }
  }
  const content = parseTextData(context, endIndex);
  return {
    type: NodeType.TEXT,
    content,
  };
}
function parseTextData(context: Context, length: number) {
  const content = context.source.slice(0, length);
  advanceBy(context, length);
  return content;
}
/**
 * 创建根节点
 * @param children
 * @returns
 */
function createRoot(children: any) {
  return {
    children,
  };
}
/**
 * 移进context
 * @param context
 * @param length
 */
function advanceBy(context: Context, length: number) {
  context.source = context.source.slice(length);
}
function isEnd(context: Context, ancestors: any[]): boolean {
  const s = context.source;
  if (s.startsWith("</")) {
    if (
      ancestors.some((ancestor) => {
        return s.startsWith(`</${ancestor.tag}>`);
      })
    ) {
      return true;
    }
  }
  return !s;
}
