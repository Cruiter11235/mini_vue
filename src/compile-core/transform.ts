import { NodeType } from "./ast";
import { helperMapName, TO_DISPLAY_STRING } from "./runtimeHelpers";

export function transform(root: any, options = {}) {
  const context = createTransformContext(root, options);
  traverseNode(root, context);
  createRootCodegen(root);
  root.helpers = [...context.helpers.keys()]; // 解构出所有的helperKey
}
/**
 * create context
 * @param root
 * @param options
 * @returns
 */
function createTransformContext(root: any, options: any): any {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: new Map(),
    helper(key: string) {
      context.helpers.set(key, 1);
    },
  };
  return context;
}
/**
 * 遍历节点
 * @param node
 * @param context
 */
function traverseNode(node: any, context: any) {
  const nodeTransforms = context.nodeTransforms;
  for (let i = 0; i < nodeTransforms.length; i++) {
    const transformFunc = nodeTransforms[i];
    transformFunc(node);
  }
  switch (node.type) {
    case NodeType.INTERPOLATION:
      context.helper(helperMapName[TO_DISPLAY_STRING]);
      break;
    case NodeType.ROOT:
    case NodeType.ELEMENT:
      traverseChildren(node, context);
      break;
    default:
      break;
  }
}
/**
 * 遍历子节点
 * @param node
 * @param context
 */
function traverseChildren(node: any, context: any) {
  const children = node.children;
  if (children) {
    for (let i = 0; i < children.length; i++) {
      const node = children[i];
      traverseNode(node, context);
    }
  }
}
/**
 * 设置生成代码的入口节点
 * @param root
 */
function createRootCodegen(root: any) {
  root.codegenNode = root.children[0];
}
