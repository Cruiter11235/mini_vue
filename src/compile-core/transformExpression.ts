import { NodeType } from "./ast";

/**
 * 这是一条transform
 * @param node
 */
export function transformExpression(node: any) {
  if (node.type === NodeType.INTERPOLATION) {
    node.content = processExpression(node.content);
  }
}

function processExpression(node: any) {
  node.content = "_ctx." + node.content;
  return node;
}
