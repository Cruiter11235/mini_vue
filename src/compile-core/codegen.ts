import { NodeType } from "./ast";
import { helperMapName, TO_DISPLAY_STRING } from "./runtimeHelpers";
import { transformExpression } from "./transformExpression";

export function generator(ast: any) {
  const context = createCodegenContext();
  const { push } = context;

  genFunctionPreamble(ast, context); // 生成前导码

  const functionName = "render";
  const args = ["_ctx", "_cache"];
  const signature = args.join(", ");
  push(`function ${functionName}(${signature}){`);
  const node = ast.codegenNode; // 代码生成入口节点
  genNode(node, context);
  push("}");
  return { code: context.code };
}
function genFunctionPreamble(ast: any, context: any) {
  const { push } = context;
  const VueBinding = "vue";
  const aliasHelper = (s: string) => `${s}:_${s}`;
  // resolve helpers

  if (ast.helpers.length > 0) {
    push(
      `const { ${ast.helpers.map(aliasHelper).join(",")} } = ${VueBinding};\n`
    );
  }
  push("return ");
}

/**
 * 创建context对象
 * @returns Context
 */
function createCodegenContext() {
  const context = {
    code: "",
    push(source: string) {
      context.code += source;
    },
    helper(key:any) {
      return `_${helperMapName[key]}`;
    },
  };
  return context;
}
/**
 * 将node.content 添加到context
 * @param node
 * @param context
 */
function genNode(node: any, context: any) {
  switch (node.type) {
    case NodeType.TEXT:
      genText(node, context);
      break;
    case NodeType.INTERPOLATION:
      genInterpolation(node, context);
      break;
    case NodeType.SIMPLE_EXPRESSION:
      genExpression(node, context);
      break;
  }
}
function genInterpolation(node: any, context: any) {
  const { push, helper } = context;
  push(helper(TO_DISPLAY_STRING));  
  push("(")
  genNode(node.content, context);
  push(")");
}

function genText(node: any, context: any) {
  const { push } = context;
  push(`'${node.content}'`);
}
function genExpression(node: any, context: any) {
  const { push } = context;
  push(`${node.content}`);
}
