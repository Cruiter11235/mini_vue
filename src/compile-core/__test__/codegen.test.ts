import { generator } from "../codegen";
import { baseParse } from "../parser";
import { transform } from "./../transform";
import { transformExpression } from "./../transformExpression";

describe("codegen", () => {
  it("string", () => {
    const ast = baseParse("hi");
    transform(ast);
    const { code } = generator(ast);
    expect(code).toMatchSnapshot();
  });
  it("interpolation", () => {
    const ast = baseParse("{{message}}");
    transform(ast, {
      nodeTransforms: [transformExpression],
    });
    const { code } = generator(ast);
    expect(code).toMatchSnapshot();
  });
  it("element", () => {
    const ast = baseParse("<div>hi,{{mini-vue}}</div>");
    transform(ast, {
      nodeTransforms: [transformExpression],
    });
    console.log(ast);
    const { code } = generator(ast);
  });
});
