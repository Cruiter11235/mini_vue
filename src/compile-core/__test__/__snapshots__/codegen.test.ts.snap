// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`codegen > interpolation 1`] = `
"const { toDisplayString:_toDisplayString } = vue;
return function render(_ctx, _cache){_toDisplayString(_ctx.message)}"
`;

exports[`codegen > string 1`] = `"return function render(_ctx, _cache){'hi'}"`;
